### ISSUE TITLE
Focal Loss for Multi-class classification

### ISSUE THREAD

**Issue by DevashishPrasad** [mathiaszinnen/focal_loss_torch#1]
Thank you for your work. I am sure this implementation of Focal Loss will work perectly fine for Binary Classification. Will it also work for multi-class classification? Or will need to tweak it a bit?

**Comment by alexander-soare** [comment]
You can try this out. Take it with a grain of salt though as I did it in a hurry

```
class CategoricalFocalLoss(nn.Module):
    def __init__(self, gamma:float=2., reduction:str='mean'):
        """
        https://arxiv.org/pdf/1708.02002.pdf
        default gamma from table 1(b)
        we could specify alpha as a per class weighting but I haven't implemented
         it yet
        """
        super().__init__()
        if reduction not in ['mean', 'none', 'sum']:
            raise NotImplementedError('Reduction {} not implemented.'.format(reduction))
        self.reduction = reduction
        self.gamma = gamma

    def forward(self, input: FloatTensor, target: LongTensor):
        """
        expect input to be (B, N) where N is number of classes
        expect target to be (B, 1) If target is B, unsqueeze
        """
        target = target.reshape(target.shape[0], -1)
        p = torch.softmax(input, -1)
        p = torch.gather(p, -1, target)
        logp = torch.log_softmax(input, -1)
        logp = torch.gather(logp, -1, target)
        fl = - 1 * (1 - p) ** self.gamma * logp
        return self._reduce(fl)

    def _reduce(self, x):
        if self.reduction == 'mean':
            return x.mean()
        elif self.reduction == 'sum':
            return x.sum()
        else:
            return x
```

**Comment by DevashishPrasad** [comment]
Thank you for the response!! @alexander-soare

**Comment by usama-axcelerate** [comment]
>         p = torch.softmax(input, -1)
>         p = torch.gather(p, -1, target)
>         logp = torch.log_softmax(input, -1)
>         logp = torch.gather(logp, -1, target)

Why you are calculating loss for only **true class** here? Isn't in multi class classification we add all classes losses?
